File: backend/app/api/personal.py
import re
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from sqlalchemy import text

from ..core.database import get_db
from ..models.models import Personal as PersonalModel
from ..schemas.personal import Personal, PersonalCreate, PersonalUpdate
from ..utils.logger import logger

router = APIRouter()


def format_phone_number(phone: str) -> str:
    """
    Format phone number to a standard format.
    Removes all non-digit characters and formats as (XXX) XXX-XXXX if 10 digits.

    Args:
        phone: Raw phone number string

    Returns:
        Formatted phone number
    """
    if not phone:
        return phone

    # Remove all non-digit characters
    digits = re.sub(r'\D', '', phone)

    # Format based on length
    if len(digits) == 10:
        return f"({digits[:3]}) {digits[3:6]}-{digits[6:]}"
    elif len(digits) == 11 and digits[0] == '1':
        # US number with country code
        return f"+1 ({digits[1:4]}) {digits[4:7]}-{digits[7:]}"
    else:
        # Return original if not a standard format
        return phone


@router.get("/personal")
async def get_personal_info(db: Session = Depends(get_db)):
    """
    Get personal information.

    Returns either the existing personal record or an empty object if none exists.
    """
    try:
        # Query to get any existing record (there should only be 0 or 1)
        query = text("""
            SELECT * FROM personal
            WHERE first_name IS NULL OR first_name IS NOT NULL
            LIMIT 1
        """)

        result = db.execute(query).first()

        if result:
            # Convert to dict and return
            return {
                "first_name": result.first_name or "",
                "last_name": result.last_name or "",
                "email": result.email or "",
                "phone": result.phone or "",
                "linkedin_url": result.linkedin_url or "",
                "github_url": result.github_url or "",
                "website_url": result.website_url or "",
                "portfolio_url": result.portfolio_url or "",
                "address_1": result.address_1 or "",
                "address_2": result.address_2 or "",
                "city": result.city or "",
                "state": result.state or "",
                "zip": result.zip or "",
                "country": result.country or "",
                "no_response_week": result.no_response_week,
	            "resume_extract_llm": result.resume_extract_llm or "",
                "job_extract_llm": result.job_extract_llm or "",
                "rewrite_llm": result.rewrite_llm or "",
                "cover_llm": result.cover_llm or "",
                "company_llm": result.company_llm or "",
                "openai_api_key": result.openai_api_key or "",
                "tinymce_api_key": result.tinymce_api_key or "",
                "convertapi_key": result.convertapi_key or "",
                "docx2html": result.docx2html or "docx-parser-converter",
                "odt2html": result.odt2html or "pandoc",
                "pdf2html": result.pdf2html or "markitdown",
                "html2docx": result.html2docx or "html4docx",
                "html2odt": result.html2odt or "pandoc",
                "html2pdf": result.html2pdf or "weasyprint"
            }
        else:
            # Return empty object
            return {
                "first_name": "",
                "last_name": "",
                "email": "",
                "phone": "",
                "linkedin_url": "",
                "github_url": "",
                "website_url": "",
                "portfolio_url": "",
                "address_1": "",
                "address_2": "",
                "city": "",
                "state": "",
                "zip": "",
                "country": "",
                "no_response_week": 4,
	            "resume_extract_llm": "gpt-4.1-mini",
                "job_extract_llm": "gpt-4.1-mini",
                "rewrite_llm": "gpt-4.1-mini",
                "cover_llm": "gpt-4.1-mini",
                "company_llm": "gpt-4.1-mini",
                "openai_api_key": "",
                "tinymce_api_key": "",
                "convertapi_key": "",
                "docx2html": "docx-parser-converter",
                "odt2html": "pandoc",
                "pdf2html": "markitdown",
                "html2docx": "html4docx",
                "html2odt": "pandoc",
                "html2pdf": "weasyprint"
            }

    except Exception as e:
        logger.error(f"Error fetching personal info", error=str(e))
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error fetching personal information: {str(e)}"
        )


@router.post("/personal", status_code=status.HTTP_200_OK)
async def save_personal_info(personal_data: PersonalCreate, db: Session = Depends(get_db)):
    """
    Save personal information.

    Creates a new record if none exists, otherwise updates the existing record.
    Validates email format and URL formats for relevant fields.
    Formats phone number to standard format.
    """
    try:
        # Format phone number if provided
        if personal_data.phone and personal_data.phone.strip():
            personal_data.phone = format_phone_number(personal_data.phone)

        # Check if a record already exists
        check_query = text("""
            SELECT first_name, last_name FROM personal
            WHERE first_name IS NULL OR first_name IS NOT NULL
            LIMIT 1
        """)

        existing = db.execute(check_query).first()

        if existing:
            # Update existing record using the composite primary key
            update_query = text("""
                UPDATE personal
                SET email = :email,
                    phone = :phone,
                    linkedin_url = :linkedin_url,
                    github_url = :github_url,
                    website_url = :website_url,
                    portfolio_url = :portfolio_url,
                    address_1 = :address_1,
                    address_2 = :address_2,
                    city = :city,
                    state = :state,
                    zip = :zip,
                    country = :country,
                    no_response_week = :no_response_week,
	                resume_extract_llm = :resume_extract_llm,
                    job_extract_llm = :job_extract_llm,
                    rewrite_llm = :rewrite_llm,
                    cover_llm = :cover_llm,
                    company_llm = :company_llm,
                    openai_api_key = :openai_api_key,
                    tinymce_api_key = :tinymce_api_key,
                    convertapi_key = :convertapi_key,
                    docx2html = :docx2html,
                    odt2html = :odt2html,
                    pdf2html = :pdf2html,
                    html2docx = :html2docx,
                    html2odt = :html2odt,
                    html2pdf = :html2pdf
                WHERE first_name = :old_first_name AND last_name = :old_last_name
            """)

            db.execute(update_query, {
                "old_first_name": existing.first_name,
                "old_last_name": existing.last_name,
                "first_name": personal_data.first_name,
                "last_name": personal_data.last_name,
                "email": personal_data.email,
                "phone": personal_data.phone,
                "linkedin_url": personal_data.linkedin_url,
                "github_url": personal_data.github_url,
                "website_url": personal_data.website_url,
                "portfolio_url": personal_data.portfolio_url,
                "address_1": personal_data.address_1,
                "address_2": personal_data.address_2,
                "city": personal_data.city,
                "state": personal_data.state,
                "zip": personal_data.zip,
                "country": personal_data.country,
                "no_response_week": personal_data.no_response_week,
	            "resume_extract_llm": personal_data.resume_extract_llm,
                "job_extract_llm": personal_data.job_extract_llm,
                "rewrite_llm": personal_data.rewrite_llm,
                "cover_llm": personal_data.cover_llm,
                "company_llm": personal_data.company_llm,
                "openai_api_key": personal_data.openai_api_key,
                "tinymce_api_key": personal_data.tinymce_api_key,
                "convertapi_key": personal_data.convertapi_key,
                "docx2html": personal_data.docx2html,
                "odt2html": personal_data.odt2html,
                "pdf2html": personal_data.pdf2html,
                "html2docx": personal_data.html2docx,
                "html2odt": personal_data.html2odt,
                "html2pdf": personal_data.html2pdf
            })
            db.commit()

            logger.info(f"Updated personal information", first_name=existing.first_name, last_name=existing.last_name)

        else:
            # Insert new record
            insert_query = text("""
                INSERT INTO personal (
                    first_name, last_name, email, phone,
                    linkedin_url, github_url, website_url, portfolio_url,
                    address_1, address_2, city, state, zip, country, no_response_week,
                    resume_extract_llm, job_extract_llm, rewrite_llm, cover_llm, company_llm,
                    openai_api_key, tinymce_api_key, convertapi_key,
                    docx2html, odt2html, pdf2html, html2docx, html2odt, html2pdf
                ) VALUES (
                    :first_name, :last_name, :email, :phone,
                    :linkedin_url, :github_url, :website_url, :portfolio_url,
                    :address_1, :address_2, :city, :state, :zip, :country, :no_response_week,
                    :resume_extract_llm, :job_extract_llm, :rewrite_llm, :cover_llm, :company_llm,
                    :openai_api_key, :tinymce_api_key, :convertapi_key,
                    :docx2html, :odt2html, :pdf2html, :html2docx, :html2odt, :html2pdf
                )
            """)

            db.execute(insert_query, {
                "first_name": personal_data.first_name,
                "last_name": personal_data.last_name,
                "email": personal_data.email,
                "phone": personal_data.phone,
                "linkedin_url": personal_data.linkedin_url,
                "github_url": personal_data.github_url,
                "website_url": personal_data.website_url,
                "portfolio_url": personal_data.portfolio_url,
                "address_1": personal_data.address_1,
                "address_2": personal_data.address_2,
                "city": personal_data.city,
                "state": personal_data.state,
                "zip": personal_data.zip,
                "country": personal_data.country,
                "no_response_week": personal_data.no_response_week,
	            "resume_extract_llm": personal_data.resume_extract_llm,
                "job_extract_llm": personal_data.job_extract_llm,
                "rewrite_llm": personal_data.rewrite_llm,
                "cover_llm": personal_data.cover_llm,
                "company_llm": personal_data.company_llm,
                "openai_api_key": personal_data.openai_api_key,
                "tinymce_api_key": personal_data.tinymce_api_key,
                "convertapi_key": personal_data.convertapi_key,
                "docx2html": personal_data.docx2html,
                "odt2html": personal_data.odt2html,
                "pdf2html": personal_data.pdf2html,
                "html2docx": personal_data.html2docx,
                "html2odt": personal_data.html2odt,
                "html2pdf": personal_data.html2pdf
            })
            db.commit()

            logger.info(f"Created new personal information record")

        return {"status": "success"}

    except ValueError as e:
        # Validation error from Pydantic
        logger.warning(f"Validation error saving personal info", error=str(e))
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        logger.error(f"Error saving personal info", error=str(e))
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error saving personal information: {str(e)}"
        )

File: frontend/src/pages/Personal/Personal.js
import React, {useState, useEffect} from 'react';
import apiService from '../../services/api';
import './Personal.css';

const Personal = () => {
    const [isEditing, setIsEditing] = useState(false);
    const [personalData, setPersonalData] = useState({
        first_name: '',
        last_name: '',
        email: '',
        phone: '',
        linkedin_url: '',
        github_url: '',
        website_url: '',
        portfolio_url: '',
        address_1: '',
        address_2: '',
        city: '',
        state: '',
        zip: '',
        country: '',
        no_response_week: 4,
        resume_extract_llm: 'gpt-4.1-mini',
        job_extract_llm: 'gpt-4.1-mini',
        rewrite_llm: 'gpt-4.1-mini',
        cover_llm: 'gpt-4.1-mini',
        openai_api_key: '',
        tinymce_api_key: '',
        convertapi_key: '',
        docx2html: 'docx-parser-converter',
        odt2html: 'pandoc',
        pdf2html: 'markitdown',
        html2docx: 'html4docx',
        html2odt: 'pandoc',
        html2pdf: 'weasyprint'
    });
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);

    // Available LLM models
    const llmModels = [
        "gpt-5.2",
        "gpt-5.2-2025-12-11",
        "gpt-5.2-chat-latest",
        "gpt-5.2-pro",
        "gpt-5.2-pro-2025-12-11",
        "gpt-5.1",
        "gpt-5.1-codex",
        "gpt-5.1-codex-max",
        "gpt-5.1-codex-mini",
        "gpt-5.1-mini",
        "gpt-5.1-nano",
        "gpt-5.1-chat-latest",
        "gpt-5.1-2025-11-13",
        "gpt-5-mini",
        "gpt-5-mini-2025-08-07",
        "gpt-5-nano",
        "gpt-5-nano-2025-08-07",
        "gpt-5-codex",
        "gpt-5",
        "gpt-5-2025-08-07",
        "gpt-5-chat-latest",
        "gpt-5-pro",
        "gpt-5-search-api",
        "gpt-5-search-api-2025-10-14",
        "gpt-5-pro-2025-10-06",
        "gpt-4.1",
        "gpt-4-1106-preview",
        "gpt-4.1-2025-04-14",
        "gpt-4.1-nano-2025-04-14",
        "gpt-4.1-mini",
        "gpt-4.1-mini-2025-04-14",
        "gpt-4.1-nano",
        "gpt-4",
        "gpt-4-0125-preview",
        "gpt-4-turbo",
        "gpt-4-turbo-preview",
        "gpt-4-turbo-2024-04-09",
        "gpt-4-0613",
        "gpt-4o-2024-08-06",
        "chatgpt-4o-latest",
        "gpt-4o",
        "gpt-4o-realtime-preview",
        "gpt-4o-2024-11-20",
        "gpt-4o-2024-05-13",
        "gpt-4o-audio-preview",
        "gpt-4o-audio-preview-2025-06-03",
        "gpt-4o-audio-preview-2024-12-17",
        "gpt-4o-audio-preview-2024-10-01",
        "gpt-4o-mini",
        "o4-mini",
        "gpt-4o-mini-2024-07-18",
        "o4-mini-2025-04-16",
        "gpt-4o-mini-audio-preview",
        "gpt-4o-mini-audio-preview-2024-12-17",
        "o4-mini-deep-research-2025-06-26",
        "o4-mini-deep-research",
        "gpt-4o-mini-realtime-preview",
        "gpt-4o-mini-realtime-preview-2024-12-17",
        "gpt-4o-mini-search-preview",
        "gpt-4o-mini-search-preview-2025-03-11",
        "gpt-4o-mini-transcribe",
        "gpt-4o-mini-tts",
        "gpt-realtime-mini",
        "gpt-realtime-mini-2025-10-06",
        "gpt-4o-realtime-preview-2025-06-03",
        "gpt-4o-realtime-preview-2024-12-17",
        "gpt-4o-realtime-preview-2024-10-01",
        "gpt-4o-search-preview",
        "gpt-4o-search-preview-2025-03-11",
        "gpt-4o-transcribe",
        "gpt-4o-transcribe-diarize",
        "gpt-3.5-turbo",
        "gpt-3.5-turbo-0125",
        "gpt-3.5-turbo-1106",
        "gpt-3.5-turbo-instruct",
        "gpt-3.5-turbo-instruct-0914",
        "gpt-3.5-turbo-16k",
        "o3",
        "o3-2025-04-16",
        "o3-mini",
        "o3-mini-2025-01-31",
        "o1",
        "o1-2024-12-17",
        "o1-pro",
        "o1-pro-2025-03-19",
        "gpt-audio",
        "gpt-audio-2025-08-28",
        "gpt-audio-mini",
        "gpt-audio-mini-2025-10-06",
        "gpt-image-1",
        "gpt-image-1-mini",
        "gpt-realtime",
        "gpt-realtime-2025-08-28",
        "babbage-002",
        "codex-mini-latest",
        "davinci-002",
        "dall-e-3",
        "dall-e-2",
        "omni-moderation-latest",
        "omni-moderation-2024-09-26",
        "sora-2",
        "sora-2-pro",
        "text-embedding-3-small",
        "text-embedding-3-large",
        "tts-1",
        "tts-1-hd-1106",
        "tts-1-hd",
        "tts-1-1106",
        "whisper-1"
    ];

    // Personal Information fields (left column)
    const personalFields = {
        first_name: 'First Name',
        last_name: 'Last Name',
        email: 'Email',
        phone: 'Phone Number',
        linkedin_url: 'LinkedIn URL',
        github_url: 'Github URL',
        website_url: 'Personal Website',
        portfolio_url: 'Portfolio URL',
        address_1: 'Street Address',
        address_2: 'Street Address 2',
        city: 'City',
        state: 'State/Province',
        zip: 'Zip Code',
        country: 'Country'
    };

    // Settings fields (right column)
    const settingsFields = {
        no_response_week: 'Auto Status Change',
        resume_extract_llm: 'Resume Extract LLM',
        job_extract_llm: 'Job Extraction LLM',
        rewrite_llm: 'Resume Rewrite LLM',
        cover_llm: 'Cover Letter LLM',
        openai_api_key: 'OpenAI API Key',
        tinymce_api_key: 'TinyMCE API Key',
        convertapi_key: 'ConvertAPI Key',
        docx2html: 'Docx To HTML',
        odt2html: 'Odt To HTML',
        pdf2html: 'PDF To HTML',
        html2docx: 'HTML To Docx',
        html2odt: 'HTML To Odt',
        html2pdf: 'HTML To PDF'
    };

    // Conversion method options
    const conversionOptions = {
        docx2html: [
            { value: 'docx-parser-converter', label: 'docx-parser-converter' },
            { value: 'convertapi', label: 'convertapi' },
            { value: 'docx2html', label: 'Docx2Html' }
        ],
        odt2html: [
            { value: 'pandoc', label: 'pandoc' }
        ],
        pdf2html: [
            { value: 'markitdown', label: 'markitdown' }
        ],
        html2docx: [
            { value: 'html4docx', label: 'html4docx' },
            { value: 'convertapi', label: 'convertapi' },
            { value: 'python-docx', label: 'Python-Docx' }
        ],
        html2odt: [
            { value: 'pandoc', label: 'pandoc' },
            { value: 'convertapi', label: 'convertapi' }
        ],
        html2pdf: [
            { value: 'weasyprint', label: 'weasyprint' },
            { value: 'convertapi', label: 'convertapi' }
        ]
    };

    useEffect(() => {
        fetchPersonalInfo();
    }, []);

    const fetchPersonalInfo = async () => {
        setLoading(true);
        setError(null);
        try {
            const data = await apiService.getPersonalInfo();
            // Apply defaults for fields that may be null/empty
            setPersonalData({
                ...data,
                no_response_week: data.no_response_week || 4,
                resume_extract_llm: data.resume_extract_llm || 'gpt-4.1-mini',
                job_extract_llm: data.job_extract_llm || 'gpt-4.1-mini',
                rewrite_llm: data.rewrite_llm || 'gpt-4.1-mini',
                cover_llm: data.cover_llm || 'gpt-4.1-mini',
                openai_api_key: data.openai_api_key || '',
                tinymce_api_key: data.tinymce_api_key || '',
                convertapi_key: data.convertapi_key || '',
                docx2html: data.docx2html || 'docx-parser-converter',
                odt2html: data.odt2html || 'pandoc',
                pdf2html: data.pdf2html || 'markitdown',
                html2docx: data.html2docx || 'html4docx',
                html2odt: data.html2odt || 'pandoc',
                html2pdf: data.html2pdf || 'weasyprint'
            });
        } catch (err) {
            setError(err.message);
            console.error('Error fetching personal info:', err);
        } finally {
            setLoading(false);
        }
    };

    const handleEdit = () => {
        setIsEditing(true);
    };

    const handleCancel = () => {
        setIsEditing(false);
        fetchPersonalInfo(); // Reset to original data
    };

    const handleChange = (e) => {
        const {name, value} = e.target;
        setPersonalData(prev => ({
            ...prev,
            [name]: value
        }));
    };

    const handleSave = async () => {
        setError(null);
        try {
            await apiService.savePersonalInfo(personalData);
            setIsEditing(false);
        } catch (err) {
            setError(err.message);
            console.error('Error saving personal info:', err);
        }
    };

    if (loading) {
        return <div className="personal-container">Loading...</div>;
    }

    const isLlmField = (fieldName) => {
        return ['resume_extract_llm', 'job_extract_llm', 'rewrite_llm', 'cover_llm'].includes(fieldName);
    };

    const isApiKeyField = (fieldName) => {
        return ['openai_api_key', 'tinymce_api_key', 'convertapi_key'].includes(fieldName);
    };

    const isConversionField = (fieldName) => {
        return ['docx2html', 'odt2html', 'pdf2html', 'html2docx', 'html2odt', 'html2pdf'].includes(fieldName);
    };

    const isWideField = (fieldName) => {
        return ['email', 'linkedin_url', 'github_url', 'website_url', 'portfolio_url', 'address_1', 'openai_api_key', 'tinymce_api_key', 'convertapi_key'].includes(fieldName);
    };

    const renderField = (fieldName, label) => {
        return (
            <div key={fieldName} className="personal-field-row">
                <label className="personal-label">
                    {label}
                </label>
                {isEditing ? (
                    <div className="personal-input-wrapper">
                        {isLlmField(fieldName) ? (
                            <select
                                name={fieldName}
                                value={personalData[fieldName] || ''}
                                onChange={handleChange}
                                className="personal-input"
                            >
                                <option value="">Select a model...</option>
                                {llmModels.map((model) => (
                                    <option key={model} value={model}>
                                        {model}
                                    </option>
                                ))}
                            </select>
                        ) : isConversionField(fieldName) ? (
                            <select
                                name={fieldName}
                                value={personalData[fieldName] || ''}
                                onChange={handleChange}
                                className="personal-input"
                            >
                                {conversionOptions[fieldName].map((option) => (
                                    <option
                                        key={option.value}
                                        value={option.value}
                                        disabled={option.value === 'convertapi' && !personalData.convertapi_key}
                                    >
                                        {option.label}
                                    </option>
                                ))}
                            </select>
                        ) : (
                            <>
                                <input
                                    type={isApiKeyField(fieldName) ? 'password' : (fieldName === 'no_response_week' ? 'number' : 'text')}
                                    name={fieldName}
                                    value={personalData[fieldName] || ''}
                                    onChange={handleChange}
                                    className={isWideField(fieldName) ? 'personal-input personal-input-wide' : 'personal-input'}
                                    min={fieldName === 'no_response_week' ? '1' : undefined}
                                    max={fieldName === 'no_response_week' ? '52' : undefined}
                                />
                                {fieldName === 'no_response_week' && (
                                    <span className="personal-input-suffix">weeks of no contact</span>
                                )}
                            </>
                        )}
                    </div>
                ) : (
                    <span className="personal-value">
                        {personalData[fieldName]
                            ? (isApiKeyField(fieldName)
                                ? '••••••••'
                                : (fieldName === 'no_response_week'
                                    ? `${personalData[fieldName]} weeks of no contact`
                                    : personalData[fieldName]))
                            : '-'}
                    </span>
                )}
            </div>
        );
    };

    return (
        <div className="personal-container">
            <div className="personal-header">
                <h1 className="personal-title">Personal Information / Settings</h1>
                {!isEditing && (
                    <button className="edit-button" onClick={handleEdit}>
                        Edit
                    </button>
                )}
            </div>

            {error && <div className="error-message">{error}</div>}

            <div className="personal-columns">
                <div className="personal-column">
                    <h2 className="column-title">Personal</h2>
                    <div className="personal-content">
                        {Object.keys(personalFields).map((fieldName) =>
                            renderField(fieldName, personalFields[fieldName])
                        )}
                    </div>
                </div>

                <div className="personal-column">
                    <h2 className="column-title">Settings</h2>
                    <div className="personal-content">
                        {Object.keys(settingsFields).map((fieldName) =>
                            renderField(fieldName, settingsFields[fieldName])
                        )}
                    </div>
                </div>
            </div>

            {isEditing && (
                <div className="personal-actions">
                    <button className="cancel-button" onClick={handleCancel}>
                        Cancel
                    </button>
                    <button className="save-button" onClick={handleSave}>
                        Save Changes
                    </button>
                </div>
            )}
        </div>
    );
};

export default Personal;
